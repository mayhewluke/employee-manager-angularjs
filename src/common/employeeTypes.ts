export enum ShiftDay {
  Monday = "Monday",
  Tuesday = "Tuesday",
  Wednesday = "Wednesday",
  Thursday = "Thursday",
  Friday = "Friday",
  Saturday = "Saturday",
  Sunday = "Sunday",
}

export interface Employee {
  employeeName: string;
  phone: string;
  shift: ShiftDay;
}
