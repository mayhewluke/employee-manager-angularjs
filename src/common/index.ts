import angular from "angular";

export const commonModule = angular.module("app.common", []).name;
