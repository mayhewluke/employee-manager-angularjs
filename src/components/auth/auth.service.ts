import angular from "angular";

import firebase from "firebase";

export class AuthService {
  constructor(private $q: ng.IQService) {}

  public logIn(
    email: string,
    password: string,
  ): ng.IPromise<firebase.auth.UserCredential> {
    return this.$q.resolve(
      firebase
        .auth()
        .signInWithEmailAndPassword(email, password)
        .catch(() =>
          firebase.auth().createUserWithEmailAndPassword(email, password),
        ),
    );
  }
}

export const authServiceModule = angular
  .module("auth.service", [])
  .service("AuthService", AuthService).name;
