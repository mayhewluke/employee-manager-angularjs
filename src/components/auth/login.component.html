<form class="login-form form-horizontal" ng-submit="$ctrl.submit($ctrl.email, $ctrl.password)">
  <div class="form-group">
    <label class="col-xs-2 control-label">Email</label>
    <div class="col-xs-10">
      <input class="form-control" name="email" id="email" type="text" ng-model="$ctrl.email" required>
    </div>
  </div>

  <div class="form-group">
    <label class="col-xs-2 control-label">Password</label>
    <div class="col-xs-10">
      <input class="form-control" name="password" id="password" type="password" ng-model="$ctrl.password" required>
    </div>
  </div>

  <div class="form-group">
    <div class="col-xs-offset-2 col-xs-10">
      <button class="btn btn-primary" type="submit">Log in</button>
      <p class="error alert alert-danger" ng-if="$ctrl.hasError">Something went wrong!</p>
    </div>
  </div>
</form>
