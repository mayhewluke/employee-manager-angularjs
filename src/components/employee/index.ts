import angular from "angular";

import { employeeServiceModule } from "./employee.service";
import listComponent from "./list.component";

export const employeeModule = angular.module("app.components.employee", [
  employeeServiceModule,
  listComponent,
]).name;

export * from "./employee.service";
