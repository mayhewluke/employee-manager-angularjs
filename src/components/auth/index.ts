import {
  HookMatchCriteria,
  TransitionHookFn,
  TransitionService,
} from "@uirouter/angularjs";
import angular from "angular";

import { authServiceModule } from "./auth.service";
import loginComponent from "./login.component";

export const authModule = angular
  .module("app.components.auth", [authServiceModule, loginComponent])
  .config(
    // tslint:disable-next-line:no-shadowed-variable
    ($transitionsProvider: TransitionService) => {
      const requireAuthCriteria: HookMatchCriteria = {
        to: state => {
          try {
            return state!.data!.requireAuth === true;
          } catch {
            return false;
          }
        },
      };
      const requireAuthHook: TransitionHookFn = transition =>
        transition
          .injector()
          .get("AuthService")
          .isAuthenticated()
          ? true
          : transition.router.stateService.target("login");
      $transitionsProvider.onBefore(requireAuthCriteria, requireAuthHook);
    },
  ).name;

export * from "./auth.service";
