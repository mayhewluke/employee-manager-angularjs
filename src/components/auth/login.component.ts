import {
  default as uiRouterModule,
  StateProvider,
  StateService,
} from "@uirouter/angularjs";
import angular, { IComponentOptions } from "angular";

import { AuthService } from "./auth.service";

import "./login.component.css";

const config: IComponentOptions = {
  bindings: {
    email: "<",
    password: "<",
  },
  controller: class LoginFormController {
    public email!: string;
    public password!: string;
    public hasError: boolean = false;

    /* @ngInject */
    public constructor(
      private $state: StateService,
      private AuthService: AuthService, // tslint:disable-line:no-shadowed-variable
    ) {}

    public submit(email: string, password: string) {
      this.AuthService.logIn(email, password)
        .then(() => {
          this.hasError = false;
          this.$state.go("employees");
        })
        .catch(() => {
          this.hasError = true;
        });
    }
  },
  // TODO add spinner when request is in progress
  template: require("./login.component.html"),
};

const name = angular
  .module("loginComponent", [uiRouterModule])
  .component("login", config)
  .config(
    /* @ngInject */
    ($stateProvider: StateProvider) => {
      $stateProvider.state("login", {
        component: "login",
        url: "/login",
      });
    },
  ).name;

export default name;
