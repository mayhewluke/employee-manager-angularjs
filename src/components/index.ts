import angular from "angular";

import { authModule } from "./auth";

export const componentsModule = angular.module("app.components", [authModule])
  .name;
