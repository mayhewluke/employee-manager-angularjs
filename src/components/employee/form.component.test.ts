import angular, { ICompileService } from "angular";
import "angular-mocks";

import { Employee, ShiftDay } from "common/employeeTypes";
import { authModule } from "components/auth";

import formComponent from "./form.component";

describe("employee form component", () => {
  const onUpdate = jest.fn();
  let render: (employee?: Employee) => JQLite;
  let scope: any;
  const testEmployee: Employee = {
    employeeName: "Taylor",
    phone: "555-5555",
    shift: ShiftDay.Monday,
  };

  beforeEach(angular.mock.module(formComponent, authModule));
  beforeEach(
    angular.mock.inject(($rootScope, $compile: ICompileService) => {
      scope = $rootScope.$new();
      render = employee => {
        scope.employee = employee;
        scope.onUpdate = onUpdate;
        const ele = $compile(
          `<employee-form employee="employee" on-update="onUpdate"></employee-form>`,
        )(scope);
        scope.$digest();
        return ele;
      };
    }),
  );

  it("renders properly", () => {
    expect(render().html()).toMatchSnapshot();
  });

  it("prepopulates the fields with the given employee's values", () => {
    const element = render(testEmployee);

    expect(element.find("input[name=name]").val()).toEqual(
      testEmployee.employeeName,
    );
    expect(element.find("input[name=phone]").val()).toEqual(testEmployee.phone);
    // Can't match on .val() because AngularJS inserts extra data into that attr
    expect(element.find("select[name=shift] option[selected]").text()).toEqual(
      ShiftDay[testEmployee.shift],
    );
  });

  describe("when changed", () => {
    it("doesn't affect the passed in employee", () => {
      const orig = angular.copy(testEmployee);
      const element = render(testEmployee);

      element
        .find("input[name=name]")
        .val("foo")
        .triggerHandler("input");
      element
        .find("input[name=phone]")
        .val("123-456-7890")
        .triggerHandler("input");
      // TODO this one does not work - figure out what handler needs to be used to make this update
      element
        .find("select[name=shift]")
        .val("Friday")
        .triggerHandler("input");

      expect(testEmployee).toEqual(orig);
    });

    it("calls the onUpdate binding with the new employee state", () => {
      const orig = {
        employeeName: "Taylor",
        phone: "555-5555",
        shift: ShiftDay.Monday,
      };
      const expected = { ...orig, phone: "98754" };
      const element = render(orig);

      element
        .find("input[name=phone]")
        .val(expected.phone)
        .triggerHandler("input");

      expect(onUpdate).toHaveBeenCalledTimes(1);
      expect(onUpdate).toHaveBeenCalledWith(expected);
    });
  });
});
