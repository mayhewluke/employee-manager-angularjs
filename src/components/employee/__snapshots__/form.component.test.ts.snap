// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`employee form component renders properly 1`] = `
"
    <label>Name
      <input name=\\"name\\" type=\\"text\\" placeholder=\\"Taylor\\" ng-model=\\"$ctrl.employee.employeeName\\" ng-change=\\"$ctrl.onUpdate($ctrl.employee)\\" class=\\"ng-pristine ng-untouched ng-valid ng-empty\\">
    </label>
    <label>Phone
      <input name=\\"phone\\" type=\\"text\\" placeholder=\\"555-555-5555\\" ng-model=\\"$ctrl.employee.phone\\" ng-change=\\"$ctrl.onUpdate($ctrl.employee)\\" class=\\"ng-pristine ng-untouched ng-valid ng-empty\\">
    </label>
    <label>Shift
      <select name=\\"shift\\" ng-model=\\"$ctrl.employee.shift\\" ng-options=\\"value as humanized for (value, humanized) in $ctrl.options\\" ng-change=\\"$ctrl.onUpdate($ctrl.employee)\\" class=\\"ng-pristine ng-untouched ng-valid ng-empty\\"><option value=\\"?\\" selected=\\"selected\\"></option><option value=\\"string:Monday\\">Monday</option><option value=\\"string:Tuesday\\">Tuesday</option><option value=\\"string:Wednesday\\">Wednesday</option><option value=\\"string:Thursday\\">Thursday</option><option value=\\"string:Friday\\">Friday</option><option value=\\"string:Saturday\\">Saturday</option><option value=\\"string:Sunday\\">Sunday</option></select>
    </label>
  "
`;
